packages/core/echo/echo-db/src/core-db/automerge-repo-replacement/repo.ts

```diff
@@ -0,0 +1,194 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { DeferredTask, sleep } from '@dxos/async';
+import {
+  type DocumentId,
+  generateAutomergeUrl,
+  parseAutomergeUrl,
+  interpretAsDocumentId,
+  type AnyDocumentId,
+} from '@dxos/automerge/automerge-repo';
+import { type Stream } from '@dxos/codec-protobuf';
+import { Resource } from '@dxos/context';
+import { invariant } from '@dxos/invariant';
+import { PublicKey } from '@dxos/keys';
+import { log } from '@dxos/log';
+import {
+  type DataService,
+  type BatchedDocumentUpdates,
+  type DocumentUpdate,
+} from '@dxos/protocols/proto/dxos/echo/service';
+
+import { DocHandleReplacement } from './doc-handle';
+
+const UPDATE_BATCH_INTERVAL = 100;
+
+export class RepoReplacement extends Resource {
+  // TODO(mykola): Change to Map<string, DocHandleReplacement<unknown>>.
+  private readonly _handles: Record<string, DocHandleReplacement<any>> = {};
+  private readonly _subscriptionId = PublicKey.random().toHex();
+  private _subscription?: Stream<BatchedDocumentUpdates> = undefined;
+
+  private readonly _docsWithPendingUpdates = new Set<DocumentId>();
+  private readonly _writeJob = new DeferredTask(this._ctx, async () => {
+    const updates: DocumentUpdate[] = [];
+
+    const docsWithPendingUpdates = Array.from(this._docsWithPendingUpdates);
+    this._docsWithPendingUpdates.clear();
+
+    for (const documentId of docsWithPendingUpdates) {
+      const handle = this._handles[documentId];
+      invariant(handle, `No handle found for documentId ${documentId}`);
+      const update = handle._getLastWriteMutation();
+      if (update) {
+        updates.push({
+          documentId,
+          mutation: update,
+        });
+      }
+    }
+
+    if (updates.length > 0) {
+      await this._dataService.write({ updates });
+      await sleep(UPDATE_BATCH_INTERVAL);
```

Comment:
implement debounce in a way that doesn't keep this task active after it has finished -- see indexer.ts